### ISSUE TITLE
Indexer can get caught in an infinite loops of NTFS junctions or symlinks

### ISSUE THREAD

**Issue by AustinWise** [Microsoft/winfile#131]
It is possible to use `mklink` and other tools to create a symlink pointing to a parent folder. If these sorts of directories are traversed naively, it is possible to get caught traversing a never-ending directory hierarchy.

winfile's indexer appears to have this problem. There are a couple of obvious solutions:
- Don't traverse past reparse points. This is easy to implement, but prevents indexing through all junctions and symlinks.
- Detect if a directory has been visited before. This is more work to implement, but it would allow indexing more directory structures.

**Comment by craigwi** [comment]
The fix in #132 ignored certain useful and common cases (e.g., OneDrive).  Please look for a different solution.

**Comment by clzls** [comment]
May limiting the depth of path to 255 works?

**Comment by AustinWise** [comment]
As @clzls suggests, I don't think the indexer will get stuck in a true infinite loop. Eventually the indexer will bump into the MAX_PATH limit.

Another way of stopping the indexer from going in loops would be to check the "inode number" of the directory and see if it matches that of any of the parent directories. Off the top of my head I don't know how to do that with WIN32, but I will investigate.

**Comment by malxau** [comment]
Some options:

- Check the reparse tag.  FindFirstFile/FindNextFile put this in dwReserved0.  If it's IO_REPARSE_TAG_MOUNT_POINT or IO_REPARSE_TAG_SYMLINK, apply the earlier logic.
- Check the name surrogate bit in the reparse tag.  0x20000000 is supposed to be set if "the file represents another named entity in the system", which both of the above do.
- Open the object and check whether the name was redirected via GetFinalPathNameByHandle (requires Vista+.)
- As far as the inode number, I think this also requires opening the file and using GetFileInformationByHandle (see nFileIndexHigh/nFileIndexLow.)  Note though that while things like NTFS have a system similar to a UNIX inode, other file systems do not, so I don't know how universally this will work.

**Comment by clzls** [comment]
I've tested the situation on my PC. When I chose `Expand &Branch`, the Winfile didn't go into infinite loops. It stopped at a certain depth.
When you try to click on the symbol link `Loop2` on the left panel, it expands normally.
When you try to click on the symbol link `Loop2` on the right panel, Winfile presents a error telling you that it's not a valid path. (And change the current drive to `L:`)
So, @craigwi  maybe we can close this issue?

![image](https://user-images.githubusercontent.com/16474386/40461248-f3962e30-5f3c-11e8-80dc-c0008a917aa0.png)
